src/Service/MagicLinks/MagicLinkService.cs

```diff
@@ -1,27 +1,103 @@
+        if (minutelyDispatchedEmailCount >= minutelyRateLimit)
+        {
+            throw new ApiException(
+                "You have reached your rate limit for magic links. Please try again later.",
+                429
+            );
+        }
+    }
 
-    public MagicLinkService(IFido2Service fido2Service, IMailProvider mailProvider)
+    private async Task ValidateQuotaAsync(MagicLinkDTO dto)
     {
-        _fido2Service = fido2Service;
-        _mailProvider = mailProvider;
+        var now = timeProvider.GetUtcNow();
+        var account = await tenantStorage.GetAccountInformation();
+        var accountAge = (now - account.CreatedAt).Duration();
+        var isFreeAccount = account.SubscriptionTier == "Free";
+
+        // Applications created less than 24 hours ago can only send magic links to the admin email address
+        if (accountAge < TimeSpan.FromHours(24) &&
+            !account.AdminEmails.Contains(dto.EmailAddress.Address, StringComparer.OrdinalIgnoreCase))
+        {
+            throw new ApiException(
+                "Because your application has been created less than 24 hours ago, " +
+                "you can only request magic links to the admin email address.",
+                403
+            );
+        }
+
+        const int maxFreeMonthlyQuota = 50;
+        const int maxProMonthlyQuota = 1000;
+        var monthlyQuota = accountAge.TotalDays switch
+        {
+            // App created <24 hours ago
+            < 1 when isFreeAccount => (int)(0.2 * maxFreeMonthlyQuota),
+            < 1 => (int)(0.2 * maxProMonthlyQuota),
+            // App created <3 days ago
+            < 3 when isFreeAccount => (int)(0.5 * maxFreeMonthlyQuota),
+            < 3 => (int)(0.5 * maxProMonthlyQuota),
+            // App created <30 days ago
+            < 30 when isFreeAccount => (int)(0.75 * maxFreeMonthlyQuota),
+            < 30 => (int)(0.75 * maxProMonthlyQuota),
+            // App created >30 days ago
+            _ when isFreeAccount => maxFreeMonthlyQuota,
+            _ => maxProMonthlyQuota
+        };
+
+        const int maxFreeMinutelyRateLimit = 50;
+        const int maxProMinutelyRateLimit = 100;
+        var minutelyRateLimit = accountAge.TotalDays switch
+        {
+            // App created <24 hours ago
+            < 1 when isFreeAccount => (int)(0.2 * maxFreeMinutelyRateLimit),
+            < 1 => (int)(0.2 * maxProMinutelyRateLimit),
+            // App created <3 days ago
+            < 3 when isFreeAccount => (int)(0.5 * maxFreeMinutelyRateLimit),
+            < 3 => (int)(0.5 * maxProMinutelyRateLimit),
+            // App created <30 days ago
+            < 30 when isFreeAccount => (int)(0.75 * maxFreeMinutelyRateLimit),
+            < 30 => (int)(0.75 * maxProMinutelyRateLimit),
+            // App created >30 days ago
+            _ when isFreeAccount => maxFreeMinutelyRateLimit,
+            _ => maxProMinutelyRateLimit
+        };
```

Comment:
@abergs does this look about right? The numbers in the original pseudocode didn't line up, so I took the liberty to tweak them myself

Reply:
Perhaps you can go with a strategy pattern to differentiate between free and paid and keep the rest.

Reply:
Get rid of this complex switch case, and use AppFeatures, see below

Reply:
Simplified the switch expression